Remove Kubernetes 1.13.x support

wilmardo opened the pull request:
# Describe what this PR does #

Removes all things kept around for 1.13 compatibility. 1.13 is deprecated now  since 1.17.0 is released.

> The Kubernetes project maintains release branches for the most recent three minor releases.

[Source](https://kubernetes.io/docs/setup/release/version-skew-policy/#supported-versions) 

Is the change backward compatible?

Yes up to 1.14

humblec commented:
LGTM.
@Madhu-1 PTAL

Madhu-1 commented:
@humblec are we planning to remove support for kube 1.13.x?

if yes @wilmardo we need to update below things

* Update readme for the same https://github.com/ceph/ceph-csi#ceph-csi-features-and-available-versions
* Remove kubernetes templates for 1.13.x  (https://github.com/ceph/ceph-csi/tree/master/deploy/rbd/kubernetes/v1.13 and https://github.com/ceph/ceph-csi/tree/master/deploy/cephfs/kubernetes/v1.13)
* Remove 1.13.x E2E changes for both cephfs and rbd

Madhu-1 commented:
@wilmardo are you planning to complete this PR? we would like to track this for  v2.0.0 as this is one of the major change in ceph-csi if you are busy I would like to take this up, please let me know your opinion on this

wilmardo commented:
I was waiting on the reply if it is desired to remove the 1.13.x stuff. 

If so I will complete this PR asap :)

Madhu-1 commented:
> I was waiting on the reply if it is desired to remove the 1.13.x stuff.
> 
> If so I will complete this PR asap :)

@wilmardo Thanks :)

Madhu-1 commented:
@wilmardo any update on this one?

wilmardo commented:
@Madhu-1 Working on it right now :)


* [ ] update PR title to remove kubernetes 1.13.x support
* [ ] Update readme for the same https://github.com/ceph/ceph-csi#ceph-csi-features-and-available-versions
* [ ] Remove kubernetes templates for 1.13.x (https://github.com/ceph/ceph-csi/tree/master/deploy/rbd/kubernetes/v1.13 and https://github.com/ceph/ceph-csi/tree/master/deploy/cephfs/kubernetes/v1.13)
* [ ] Remove 1.13.x E2E changes for both cephfs and rbd

Madhu-1 commented:
> @Madhu-1 Working on it right now :)
> 
> * [x]  update PR title to remove kubernetes 1.13.x support
> * [x]  Update readme for the same https://github.com/ceph/ceph-csi#ceph-csi-features-and-available-versions
> * [x]  Remove kubernetes templates for 1.13.x (https://github.com/ceph/ceph-csi/tree/master/deploy/rbd/kubernetes/v1.13 and https://github.com/ceph/ceph-csi/tree/master/deploy/cephfs/kubernetes/v1.13)
> * [ ]  Remove 1.13.x E2E changes for both cephfs and rbd

Thanks, @wilmardo for the update :)

wilmardo commented:
@Madhu-1 Ready for review! 🎉

nixpanic commented on README.md:
```diff
@@ -61,16 +61,16 @@ NOTE:
 
 | Plugin | Features                                                  | Feature Status | CSI Driver Version | CSI Spec Version | Ceph Cluster Version | Kubernetes Version |
 | ------ | --------------------------------------------------------- | -------------- | ------------------ | ---------------- | -------------------- | ------------------ |
-| RBD    | Dynamically provision, de-provision Block mode RWO volume | GA             | >=v1.0.0           | >=v1.0.0         | >= Mimic             | >= v1.13.0         |
```
I think the documentation can stay at v1.13.0. But there should be a mentioning of the minimal Kubernetes version a little above this section.

Madhu-1 replied:
we have removed all the 1.13 templates so we dont need to mention about 1.13 anymore, and we cant use 1.14+ templates with kube 1.13

wilmardo replied:
Can do but when do we decide to update the minimal Kubernetes version in the table? Since the 1.13 e2e test are removed there is no indication a 1.13 breaking change is introduced.

Would something like this suffice?
```markdown
NOTE:

- **`csiv0.3`** is deprecated with release of **`csi v1.1.0`**
- The helm charts **require** Kubernetes 1.14+
- The minimum tested Kubernetes version is 1.14 since release v2.0.0
```

nixpanic replied:
Keeping the older versions of when a feature was introduced seems useful to me. It serves as a history and confirmation that the functionality is tested and likely deployed in production since ... version.

Madhu-1 replied:
if we mention that we support kube 1.13 as we are removing the templates, what if the user wants to deploy ceph-csi v2.0.0 on kube 1.13?

wilmardo replied:
I my humble opinion it is good to remove the 1.13 support since there have been some issues in the past due the sheer number of files to be edited for a change. If I need to edit the provisioner deployment in the current state I need to change:

* deploy/cephfs/1.13/csi-cephfsplugin-provisioner.yaml
* deploy/cephfs/1.14+/csi-cephfsplugin-provisioner.yaml
* deploy/rbd/1.13/csi-rbdplugin-provisioner.yaml
* deploy/rbd/1.14+/csi-rbdplugin-provisioner.yaml
* charts/ceph-csi-cephfs/provisioner-deployments.yaml
* charts/ceph-csi-cephfs/provisioner-statefulset.yaml

When a change is needed in the provisioner and the nodeplugin the files to edit is doubled. A little oversight and one of the files is forgotten.

What are the advantages of supporting a scenario the upstream wouldn't support?

nixpanic replied:
I think I read the table a little different. "Kubernetes Version" means the version when this feature was introduced, not necessarily still supported. Mentioning in the "Support Matrix" paragraph that kubernetes 1.14 or newer is required since ceph-csi 2.0.0 has my preference.

I am not strongly attached to this though, so would not strongly object to updating >=v1.13 to >=1.14 either.

nixpanic commented on charts/ceph-csi-cephfs/templates/provisioner-deployment.yaml:
```diff
@@ -1,4 +1,3 @@
-{{- if semverCompare ">=1.14" .Capabilities.KubeVersion.GitVersion -}}
```
Should there be a check for <=1.13 and have it fail with some useful error?

Madhu-1 replied:
this is not needed in the doc we are mentioning we support kube 1.13+

nixpanic commented on docs/deploy-cephfs.md:
```diff
@@ -117,9 +117,7 @@ for a zero-sized volume means no quota attribute will be set.
 
 Requires Kubernetes 1.13+
```
This would now be 1.14+ ?

wilmardo replied:
Good catch! Updated :)

nixpanic commented on e2e/README.md:
```diff
@@ -46,7 +46,7 @@ following environment variables can be exported to customize kubernetes deployme
 | ------------------ | ------------------------------------------------ | ------------------------------------------------------------------ |
 | MINIKUBE_VERSION   | minikube version to install                      | latest                                                             |
-| KUBE_VERSION       | kubernetes version to install                    | v1.13.0                                                            |
+| KUBE_VERSION       | kubernetes version to install                    | v1.14.0                                                            |
```
`scripts/minikube.sh` actually uses v1.14.2 already by default.

wilmardo replied:
Updated and bumped the `scripts/minikube.sh` default to 1.14.10 to be in sync with TravisCI

Madhu-1 commented on README.md:
```diff
@@ -61,16 +61,16 @@ NOTE:
+|        | Creating and deleting snapshot                            | Alpha          | >=v1.0.0           | >=v1.0.0         | >= Mimic             | >= v1.14.0         |
+|        | Provision volume from snapshot                            | Alpha          | >=v1.0.0           | >=v1.0.0         | >= Mimic             | >= v1.14.0         |
 |        | Provision volume from another volume                      | -              | -                  | -                | -                    | -                  |
 |        | Resize volume                                             | -              | -                  | -                | -                    | -                  |
```
not as part of this PR, @humblec we need to update Readme for resize

wilmardo commented:
@Madhu-1 @nixpanic What is the consensus about Kubernetes 1.13 support?

Madhu-1 commented:
> @Madhu-1 @nixpanic What is the consensus about Kubernetes 1.13 support?

I suggest to remove 1.13 from readme also , let @humblec  and @nixpanic confirm

nixpanic commented:
On Thu, Jan 16, 2020 at 03:12:49AM -0800, Madhu Rajanna wrote:
> > @Madhu-1 @nixpanic What is the consensus about Kubernetes 1.13 support?
> 
> I suggest to remove 1.13 from readme also , let @humblec  and @nixpanic confirm

I have stated my preference before, but do not mind if 1.13 is
completely removed from the readme if others think that is clearer.

wilmardo commented:
@Madhu-1 Thanks for the pointers, updated again!

As discussed on Slack, removed the versioned deploy directory. No more `deploy/1.14+` just `deploy/` now.
Refactored the E2E test, removed the functions to determine the correct directory and removed the `updateCephfsDirPath` since the `cephfsDirPath` is correct again :)

Madhu-1 commented:
CI is failing :(

wilmardo commented:
:'(
Will look again after dinner, I seem to have missed some function calls, sorry!

wilmardo commented:
@Madhu-1 CI completes now :) Ready for review again!

Madhu-1 commented on docs/deploy-cephfs.md:
```diff
@@ -115,11 +115,9 @@ for a zero-sized volume means no quota attribute will be set.
-if your cluster version is 1.13.x please use [cephfs v1.13
-templates](../deploy/cephfs/kubernetes/v1.13) or else use [cephfs v1.14+
-templates](../deploy/cephfs/kubernetes/v1.14+)
+Use the [cephfs v1.14+ templates](../deploy/cephfs/kubernetes)
```
change `cephfs v1.14+ templates` to `cephfs templates`

Madhu-1 commented on docs/deploy-rbd.md:
```diff
@@ -71,11 +71,9 @@ is required for provisioning new RBD images.
-if your cluster version is 1.13.x please use [rbd v1.13
-templates](../deploy/rbd/kubernetes/v1.13) or else use [rbd v1.14+
-templates](../deploy/rbd/kubernetes/v1.14+)
+Use the [rbd v1.14+ templates](../deploy/rbd/kubernetes)
```
same here also change `rbd templates` instead of `v1.14+ templates`

Madhu-1 commented on e2e/cephfs.go:
```diff
@@ -89,14 +82,8 @@ var _ = Describe("cephfs", func() {
 			appPath := cephfsExamplePath + "pod.yaml"
 
 			By("checking provisioner statefulset/deployment is running")
```
checking provisioner deployment is completed?

Madhu-1 commented on e2e/rbd.go:
```diff
@@ -99,14 +92,8 @@ var _ = Describe("RBD", func() {
 			// snapshotPath := rbdExamplePath + "snapshot.yaml"
 
 			By("checking provisioner statefulset/deployment is running")
```
`checking deployment is completed`?

wilmardo commented:
@Madhu-1 Addressed the last nits. I searched for statefulset and 1.14+ and should have got all the references now.
Squashed to one commit :)

Madhu-1 commented:
@humblec @nixpanic PTAL

humblec commented:
Thanks all!
